pipelined/bevy_render2/src/render_phase/draw.rs

```diff
@@ -1,59 +1,161 @@
 use crate::render_phase::TrackedRenderPass;
-use bevy_ecs::{entity::Entity, world::World};
+use bevy_app::App;
+use bevy_ecs::{
+    all_tuples,
+    entity::Entity,
+    system::{ReadOnlySystemParamFetch, SystemParam, SystemParamItem, SystemState},
+    world::World,
+};
 use bevy_utils::HashMap;
 use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
 use std::{any::TypeId, fmt::Debug, hash::Hash};
 
-// TODO: should this be generic on "drawn thing"? would provide more flexibility and  explicitness
-// instead of hard coded draw key and sort key
-pub trait Draw: Send + Sync + 'static {
-    fn draw<'w, 's>(
-        &'s mut self,
+pub trait Draw<P: PhaseItem>: Send + Sync + 'static {
+    fn draw<'w>(
+        &mut self,
         world: &'w World,
         pass: &mut TrackedRenderPass<'w>,
         view: Entity,
-        draw_key: usize,
-        sort_key: usize,
+        item: &P,
     );
 }
 
+pub trait PhaseItem: Send + Sync + 'static {
+    type SortKey: Ord;
+    fn sort_key(&self) -> Self::SortKey;
+    fn draw_function(&self) -> DrawFunctionId;
+}
+
+// TODO: make this generic?
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
 pub struct DrawFunctionId(usize);
 
-#[derive(Default)]
-pub struct DrawFunctionsInternal {
-    pub draw_functions: Vec<Box<dyn Draw>>,
+pub struct DrawFunctionsInternal<D: PhaseItem> {
```

Comment:
`PhaseItem` should probably have one favoured generic type parameter identifier: `D` and `P` are used in this file once each, while every other occurence is `I`.

Reply:
Ah yeah good call. These types were refactored / renamed a couple of times, and apparently I left a wasteland of old generic type identifiers :)